### ISSUE TITLE
no injection of env-entry if env-entry-value is whitespace only or missing

### ISSUE THREAD

**Issue by sf-bit** [eclipse/jetty.project#4644]
In jetty-9.4.27.v20200227 (but presumably in all versions) values for environment entries that are defined in WEB-INF/jetty-env.xml (or similarly) are not injected if env-entry-value in WEB-INF/web.xml is whitespace only or missing. This does not depend on annotations, I only used injection-target in web.xml.

Hubspot does not let me upload the example webapp I prepared so I will only post excerpts:

From src/main/webapp/WEB-INF/web.xml:

```
  <env-entry>
    <env-entry-name>empty</env-entry-name>
    <env-entry-type>java.lang.String</env-entry-type>
    <env-entry-value></env-entry-value>
    <injection-target>
      <injection-target-class>webapp.Listener</injection-target-class>
      <injection-target-name>empty</injection-target-name>
    </injection-target>
  </env-entry>

```
From src/main/webapp/WEB-INF/jetty-env.xml:

```
  <New class="org.eclipse.jetty.plus.jndi.EnvEntry">
    <Arg>empty</Arg>
    <Arg type="java.lang.String">new value for empty</Arg>
    <Arg type="boolean">true</Arg>
  </New>
```

From src/main/java/webapp/Listener.java:

```
  private String empty;

  public void setEmpty (final String empty) {
    this.empty = empty;
  }

  public void contextInitialized (final ServletContextEvent servletContextEvent) {
    final ServletContext ctx = servletContextEvent.getServletContext ();
    ctx.log ("empty=" + empty);
  }

```
Output from running

java -jar start.jar etc/jetty-plus.xml

in unpacked https://repo1.maven.org/maven2/org/eclipse/jetty/jetty-distribution/9.4.27.v20200227/jetty-distribution-9.4.27.v20200227.zip:

```
2020-03-06 11:23:37.893:WARN:oejpw.PlusDescriptorProcessor:main: No value for env-entry-name empty
...
2020-03-06 11:23:38.148:INFO:oejshC.webapp:main: empty=null
```

**Comment by sf-bit** [comment]
The culprit seems to be in https://github.com/eclipse/jetty.project/blob/jetty-9.4.27.v20200227/jetty-plus/src/main/java/org/eclipse/jetty/plus/webapp/PlusDescriptorProcessor.java:

```
   117      public void visitEnvEntry(WebAppContext context, Descriptor descriptor, XmlParser.N
ode node)
   118          throws Exception
   119      {
   120          String name = node.getString("env-entry-name", false, true);
   121          String type = node.getString("env-entry-type", false, true);
   122          String valueStr = node.getString("env-entry-value", false, true);
   123  
   124          //if there's no value there's no point in making a jndi entry
   125          //nor processing injection entries
   126          if (valueStr == null || valueStr.equals(""))
   127          {
   128              LOG.warn("No value for env-entry-name " + name);
   129              return;
   130          }

```

**Comment by joakime** [comment]
That logic has been there since the beginning of Jetty's support for JNDI.
Why do you think this is invalid?

**Comment by sf-bit** [comment]
It's a violation of the spec, e.g. https://download.oracle.com/otn-pub/jcp/servlet-3.0-fr-eval-oth-JSpec/servlet-3_0-final-spec.pdf, page 158f: "The specified resource will be injected into the target during initialization of the class by either calling the set method for the target property or by setting a value into the name filed."

**Comment by sf-bit** [comment]
Additionally, from https://download.oracle.com/otn-pub/jcp/ejb-3.1-fr-eval-oth-JSpec/ejb-3_1-fr-spec.pdf, "16.4.1.3 Declaration of Simple Environment Entries in the Deployment Descriptor", page 436-440: "[...] If  the  Bean  Provider  provides  a  value  for  an  environment  entry  using  the  env-entry-value  element, the value can be changed later by the Application Assembler or Deployer. [...] It is often convenient to declare a field as an injection target, but to specify a default value in the code, [...] To support this case, **the container must** only **inject a value for the environment entry if the** application assembler  or  **deployer  has  specified  a  value  to  override  the  default  value**.  **The  env-entry-value element in the deployment descriptor is optional** when an injection target is specified. If the element is not specified, no value will be injected. [...]" (emphasis added by me).

**Comment by janbartel** [comment]
From the schema javaee-8.xsd for env-entry:

>  If a value is not specified and injection is requested, no injection will occur and no entry of the specified name will be created. 

So, if jetty-env.xml did not exist - or the 3rd argument of the EnvEntry was `false` instead of `true` -  I'd argue that the code is currently correct.

However, as jetty-env.xml allows you to specify values that will _override_ those in web.xml (that 3rd argument!), then it does seem reasonable that the injection would be created and use the value from jetty-env.xml. I'll take a look at the code and see what can be done.

**Comment by janbartel** [comment]
This was merged for 9.4.28.
